src/array.rs

```diff
@@ -183,6 +187,125 @@ impl<'a, A: Clone> TreeLike for Partition<'a, A> {
+///
+/// ## Example
+///
+/// Take a root node with children `a`, `b`, `c`.
+///
+/// ```text
+///    .
+///  / | \
+/// a  b  c
+/// ```
+///
+/// The 3-ary root is converted into two 2-ary nodes. The normal nodes `a`, `b`, `c` stay preserved.
+///
+/// ```text
+///     .
+///    / \
+///   .   c
+///  / \
+/// a   b
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum BinaryTree<T> {
+    /// Normal node (with at most two children).
+    ///
+    /// Must be constructed via [`BinaryTree::from_tree()`].
+    Normal(T),
+    /// Array of at least two children.
+    ///
+    /// Must be constructed via [`BinaryTree::from_array()`].
+    Array(Arc<[T]>, Range<usize>),
+}
+
+impl<T> BinaryTree<T> {
+    /// Access the content of a normal node.
+    pub fn as_normal(&self) -> Option<&T> {
+        match self {
+            Self::Normal(tree) => Some(tree),
+            _ => None,
+        }
+    }
+}
+
+impl<T: TreeLike> BinaryTree<T> {
+    /// Convert a tree into a binary tree.
+    ///
+    /// ## Panics
+    ///
+    /// `<T as TreeLike>::as_node` returns `Tree::Nary` with an empty array.
+    pub fn from_tree(tree: T) -> Self {
+        match tree.as_node() {
+            Tree::Nary(array) => {
+                let range = 0..array.len();
+                Self::from_array(array, range)
+            }
+            _ => Self::Normal(tree),
+        }
+    }
+
+    /// Convert a range of an array into a binary tree.
+    ///
+    /// ## Panics
+    ///
+    /// Array is empty. Range is empty.
+    ///
+    /// `<T as TreeLike>::as_node` returns `Tree::Nary` with an empty array.
+    pub fn from_array(array: Arc<[T]>, range: Range<usize>) -> Self {
+        assert!(!array.is_empty(), "Arrays must be nonempty");
+        match range.len() {
+            0 => panic!("Range must be nonempty"),
+            1 => Self::from_tree(array[range.start].clone()),
+            _ => Self::Array(array, range),
+        }
+    }
+}
+
+impl<T: TreeLike> TreeLike for BinaryTree<T> {
+    /// # Panics
+    ///
+    /// `<T as TreeLike>::as_node` returns `Tree::Nary` with an empty array.
+    fn as_node(&self) -> Tree<Self> {
+        match self {
+            Self::Normal(tree) => match tree.as_node() {
+                Tree::Nullary => Tree::Nullary,
+                Tree::Unary(l) => Tree::Unary(Self::from_tree(l)),
+                Tree::Binary(l, r) => Tree::Binary(Self::from_tree(l), Self::from_tree(r)),
+                Tree::Nary(array) => {
+                    let range = 0..array.len();
+                    Self::from_array(array, range).as_node()
+                }
+            },
+            Self::Array(array, range) => {
+                debug_assert!(2 <= range.len());
+                let n = range.len();
+                let next_pow2 = n.next_power_of_two();
+                let half = next_pow2 / 2;
+                let left = Self::from_array(array.clone(), range.start..range.start + half);
```

Comment:
In 61d8b2393ea2d0690e11e38c2b10fcd58b9e1efe:

as a style nit, I prefer calling `Arc::clone` explicitly rather than `.clone()` so it's obvious that we're just doing a refcount increment and not a "real" clone.